Communication Services - Rooms APIspecification

elavarasidc opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new service onboarding
      - [] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month. 06/30/2022
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month. 06/07/2022
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following appy to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s) 
  - Adding a new API version
    -[x] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

elavarasidc committed:
Communication Services - Rooms API Specification

elavarasidc committed:
fixed autorest validation flags

elavarasidc committed:
Fixed validations from example validator and schema validator

elavarasidc committed:
Removed unused models

ramneettung-msft committed:
Adding in API specificatin changes for idempotency. (#18717)
Roooms Spec
* Adding in API specificatin changes for idempotency.

Adding idemptoency headers to the API specification.

* Adjusting the descriptions to be more representative.

* Adding custom words file.

* Running prettier.

* Adding readme.

* Adjusting files.

* Adjusting readme file.

* Fixing unreferenced files.

* Adding a custom word for fixdate.

elavarasidc committed:
validation log

elavarasidc committed:
Update Rooms spec to accommodate the latest Rooms API changes.

mentat9 commented:
Dataplane APIs don't need ARM review: removing incorrect labels.

JeffreyRichter reviewed (commented):
- POST /rooms
  - We prefer PUT or PATCH /rooms/{roomId} to POST
- Do not put any non-2xx status codes in the swagger. Just have a default and refer to the standard Azure error response type.
- POST /rooms/{roomId}/participants/:add
  - No / before the colon; same for update & remove
  -  I wonder if you could replace the 3 POST add/update/remove operations with a single PATCH /rooms/{roomId}/participants or even just using the PATCH /rooms/{roomId} you already have
 - If GET /rooms/{roomId} already returns the participants, why do you need GET /rooms/{roomId}/participants at all?

elavarasidc committed:
Prettier check and fixed formatting as required by prettier validation.

elavarasidc committed:
Fix Spell check flag for IMF-fixdate

weidongxu-microsoft commented on package-lock.json:
```diff
@@ -3580,4 +3580,4 @@
   }
-}
\ No newline at end of file
+}
```
this file should not be in PR.

elavarasidc replied:
Removed package-lock.json from the PR.

weidongxu-microsoft commented on specification/communication/data-plane/Rooms/preview/validate_log_20220413_221737.log:
```diff
@@ -0,0 +1,23 @@
+2022-04-14T05:17:37.968Z - silly: Initializing OpenApiDiff class
```
do we need this file?

elavarasidc replied:
Not required. Removed validate log file from the PR.

weidongxu-microsoft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,846 @@
+            "required": true
+          }
+        ],
+        "responses": {
```
missing "default", see https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors

and `x-ms-error-code` header is recommended in all error responses.

elavarasidc replied:
Default error response added with header x-ms-error-code in all error response on the specification.
@weidongxu-microsoft I amlooking for API implementation refences for x-ms-error-code on Dot Net code bases. Any pointers.

weidongxu-microsoft replied:
I am not familiar with .NET SDK. By implementation references, do you mean hand-written code or generated code (from autorest)?

I think SDK will also expose the `x-ms-error-code` in response, but at that time, the JSON in response likely already be de-serialized.
The advantage of `x-ms-error-code` would be for REST API user, that if they see 4xx, they can check the header and see whether it is recoverable (without inspecting the JSON in body).

elavarasidc replied:
I was looking for backend service implementation references for services written in .NET.
On further search, was able to find a reference under communication services itself. Email services seems to have implemented x-ms-error-code header response for error responses.
https://skype.visualstudio.com/DefaultCollection/SCC/_git/ic3_spool_cosine-dep-spool.  I will refer to this for backed implementation. 
The Spec is updated as per your recommendation @weidongxu-microsoft

weidongxu-microsoft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,846 @@
+        "summary": "Update a room with given changes.",
+        "operationId": "Rooms_UpdateRoom",
+        "consumes": [
+          "application/json"
```
PATCH should support `application/merge-patch+json`

elavarasidc replied:
Addressed

weidongxu-microsoft replied:
Great. Just make sure backend expects it and can process the merge patch.

elavarasidc replied:
Yes @weidongxu-microsoft updating backed in parallel and making sure the changes can be processed fine on the back end.

weidongxu-microsoft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,846 @@
+          "200": {
+            "description": "Request successful. The action returns the participants of a room.",
+            "schema": {
+              "$ref": "#/definitions/ParticipantsCollection"
```
Do we know maximum number of the participants? Do we need to support pagination?

elavarasidc replied:
@radubulboaca Adding @radubulboaca to comment here based on what we foresee as we are targeting Public preview in Q2 2022.

radubulboaca replied:
Thank you for looking into this @weidongxu-microsoft.

For public preview, Rooms service has a soft limit of 350 participants. The main customer scenario targeted for this service are virtual visits (e.g., doctor appointments, financial advisors, classrooms), which inherently have relatively few participants. Because of this, the team made a product decision to not include pagination at this point. As usage patterns change and we start seeing demand for large number of room participants, we plan to add pagination into the API. 

Also notable in this context is the concept of 'open room', which is also present in this API. Marking the room as 'open' allows users to join the room call when they haven't been explicitly added as participants to the room via this API (assuming the users have the proper authentication tokens). As the number of call attendees increases, managing and updating the roster naturally starts to become more cumbersome for customers (regardless of pagination). Because of this natural tendency, we expect users of large rooms to default to using 'open' rooms, which relives the need for pagination.

prprabhu-ms replied:
I will note that adding pagination to a GA'ed API is a breaking change. I'd recommend adding pagination from the start for all APIs that return collections that aren't physically bounded by design (an example would be something that returns the names of countries in the world where a service is supported).

prprabhu-ms replied:
Can we follow up on the decision around pagination internally?

I am not yet convinced that this service should GA without pagination. The guidelines require pagination before GA because it is often needed after the fact and is a breaking change. We should only GA without pagination if we're saying we will limit the number of participants to a hard limit (of less than, say 350) for the next 3 years.

elavarasidc replied:
captured pagination to discussions to https://skype.visualstudio.com/SPOOL/_workitems/edit/2930042 to be discussed and worked on incrementally before GA.

weidongxu-microsoft reviewed (commented):
A few comments after quick read.

I assume we have a meeting this week?

elavarasidc committed:
Prettier check - fix formatting on examples

elavarasidc committed:
Fixing review comments:
removing package-lock.json, validate_log file
Examplese updated to remove error codes to match spec update as per review comment

elavarasidc committed:
Fix Model validation failure.

elavarasidc commented:
> I assume we have a meeting this week?

We are meeting next week on the 16th June at 4PM PST 
[Communication - Rooms] REST API Review

elavarasidc commented:
> * POST /rooms
>   * We prefer PUT or PATCH /rooms/{roomId} to POST
[Response] The update room end point /rooms/{roomId} is a PATCH end point. 
Please find link the end point in case this was not obvious.
https://github.com/Azure/azure-rest-api-specs/blob/2e0c550739a9151c959463559ed2c8c8f0230817/specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json#L125


> * Do not put any non-2xx status codes in the swagger. Just have a default and refer to the standard Azure error response type.
[Response] Addressed.

> * POST /rooms/{roomId}/participants/:add
>   * No / before the colon; same for update & remove
[Response] Addressed for add, update and remove participants end point

>   * I wonder if you could replace the 3 POST add/update/remove operations with a single PATCH /rooms/{roomId}/participants or even just using the PATCH /rooms/{roomId} you already have
[Response]
The proposal is to move away from a comprehensive 'update' API using JSON merge patch, into discrete APIs for addition and removal of participants:
(1) Provide a consistent experience with similar APIs in other ACS SDKs to handle participants (example Chat[ Add Chat Participants - REST API (Azure Communication) | Microsoft Docs](https://docs.microsoft.com/en-us/rest/api/communication/chat/chat-thread/add-chat-participants))
(2) The private preview version of Rooms service API made use of the JSON merge patch to update the Participants in a Room. However, we have found technical issues when expanding the functionality into various language SDKs. Open issue: [Board Review: Support for JSON Merge Patch · Issue #3520 · Azure/azure-](https://github.com/Azure/azure-sdk/issues/3520)[sdk](https://github.com/Azure/azure-sdk/issues/3520)[ (github.com) ](https://github.com/Azure/azure-sdk/issues/3520)

Proposed changes: [ACS Rooms - Technical solution to Add Remove and Update participants.docx](https://microsoft-my.sharepoint.com/:w:/p/elavarasid/ET9Y1GdBSX1EvGGkbcyYzSIB43LV0AgSaY-SvymMmnl-cA?e=zCrKi1&isSPOFile=1)
[Swagger UI (microsoft.com)](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fglobal.rooms.prod.communication.microsoft.com%2Fswagger%2Findex.html&data=05%7C01%7Cjorgegarc%40microsoft.com%7C40124ee8c21048e80e3008da1ce5e8fb%7C72f988bf86f141af91ab2d7cd011db47%7C0%7C0%7C637854071029941333%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=CJvzfZGbOSiGgy0bRsP%2BNJHlfl1VQ2be8ct1xzqouGA%3D&reserved=0)

> * If GET /rooms/{roomId} already returns the participants, why do you need GET /rooms/{roomId}/participants at all?
[Response] 
Expecting developers to use the GET participants end point, more frequent, before removing / updating participants.  
At this point in time rooms has only few attributes. when this grows in the future, these additional room attributes will not be necessary.
Also, other communication services such as chat has end point to list participants. Having Get participants end point will give us a similar developer experience across ACS services.

elavarasidc committed:
Removed references to Phone Number Identifier and Team Users as they are not supported
by room yet

ramneettung-msft committed:
Adding RoomOpen flag for ACS Rooms. (#19339)

elavarasidc committed:
Updated Spec to refer to Azure communications common communication user identifier

elavarasidc committed:
Adding teams and phone number communication identifier for dotnetsdk communicationidentifierSerializer code generation

elavarasidc committed:
Adding Commmunication Identifier Model

elavarasidc committed:
Updated description of CommunicationIdentifierModel

elavarasidc committed:
Fix update participants end point

prprabhu-ms commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "The participants were successfully added.",
```
nit: Document the returned `RoomModel` as in other APIs.

Also, should these `participants:<verb>` APIs return a `RoomModel`, or just the list of participants? It's a bit weird that the response for adding a participant includes information like room validity.

elavarasidc replied:
nit addressed. 
Retaining current behavior to return RoomModel for any update to a room, including update to room participants to provide a complete view to the customer how their room looks like after the action across the rooms API.

prprabhu-ms replied:
I'd recommend against this confusion between models in your API.

As a precedent, see the chat service's participants:add - https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Chat/stable/2021-09-07/communicationserviceschat.json#L1527

It returns an object that contains information about the add operation itself, but no information about even all the participants.

This is inline with CRUD operations: Create operation returns the created model, Read of course returns it as well, but Update and Delete do not.

Here's one reason why returning the RoomModel here might be bad: Let's say you want to implement `filter` for GET on `room/<id>`. Now the `participants:add` becomes another method that also returns the same RoomModel, but does not implement `filter`. This will cause confusion in the API surface, and might actually be bad if the `filter` is needed for performance reasons.

Another performance argument: Contoso decides to add 200 participants one by one with `participants:add`. Then we end up returning the room back each time, with increasing number of participants. Effectively, our cumuilative response size will be O(N^2) in the number of participants.

elavarasidc replied:
@prprabhu-ms Addressed. The response end points of Participants action end points has been updated to return participants collection instead of a room.

prprabhu-ms commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+        },
+        "role": {
+          "description": "Role Name.",
+          "type": "string"
```
This is essentially a stringly-typed enum, right? What are the permissible values?

Can you use an extensible enum here? https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#use-extensible-enums

elavarasidc replied:
Addressed: changed role from string to extensible enum.

prprabhu-ms commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+        }
+      }
+    },
+    "CommunicationIdentifierModel": {
```
I'm a bit surprised by this definition.
If I understand correctly, this is trying to keep the data shape of `CommunicaitonUserIdentifierModel` from the communication/common package, but restricting the enum to just the ACS user.

Is there precedent for this? If we need this enum-subset type here (and presumably elsewhere), I think we should still define it in the communication/common package.

petrsvihlik replied:
Yeah, I was also surprised by this. But apparently, there is a precedent - see e.g. chat:
https://docs.microsoft.com/en-us/rest/api/communication/chat/chat/create-chat-thread

petrsvihlik replied:
Aha...the `rawId` is optional
> "Raw Id of the identifier. Optional in requests, required in responses."

https://github.com/Azure/azure-rest-api-specs/blob/03a43aa61a1e02d84c902b2c2ed1bd341833dcfa/specification/communication/data-plane/Common/stable/2021-03-07/common.json#L73-L94

petrsvihlik replied:
@prprabhu-ms 
> If I understand correctly, this is trying to keep the data shape of CommunicaitonUserIdentifierModel from the communication/common package, but restricting the enum to just the ACS user.

It's in the shape of CommunicationIdentifier, not CommunicationUserIdentifier...so it doesn't restrict it only to ACS users.

prprabhu-ms replied:
My comment was about the *definition* of this model here.

Chat just references the common model: https://github.com/Azure/azure-rest-api-specs/blob/03a43aa61a1e02d84c902b2c2ed1bd341833dcfa/specification/communication/data-plane/Chat/stable/2021-09-07/communicationserviceschat.json#L1200

petrsvihlik replied:
They reference the common model on the line 522 and 527. They're just wrapping it in a parent type that also has a role attached to it.

elavarasidc replied:
Great question. Let me explain why we are not referring to the common CommunicationIdentifierModel.
At this point in time Rooms do not support Microsoft Teams User Identifier and Phone number identifier. We do have this in pipeline. 
In order to refer only to the Communication User Identifier which Rooms support right now, we have defined CommunicationIdentifierModel at our spec level rather than using the common one which includes all 3 identifiers and referring to CommunicationUserIdentifierModel from the communication common. 
Hope that clarifies.

petrsvihlik replied:
I get your point @prprabhu-ms, you're right.
@elavarasidc Yes, all clear. Just pehaps one note - make sure that your specification of the `CommunicationIdentifierModel` never diverges from the common-level spec so that you'll be always able to make the transition back.

prprabhu-ms replied:
Will they be able to make the transition though? That's my main worry.

`CommunicationIdentifierModel` is a polymorphic type. I'm not sure if adding more variants to a polymorphic type is considered a breaking change. That's a question for the API stewardship council.

elavarasidc replied:
> I get your point @prprabhu-ms, you're right. @elavarasidc Yes, all clear. Just perhaps one note - make sure that your specification of the `CommunicationIdentifierModel` never diverges from the common-level spec so that you'll be always able to make the transition back.

@petrsvihlik yes, we the plan is to make sure we align with the common CommunicationIdentifierModel.

prprabhu-ms commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+          "type": "string",
+          "example": "2021-09-08T15:55:41Z"
+        },
+        "roomOpen": {
```
As a general principle, avoid boolean flags in API models.
It's worth thinking about whether this should be an extensible enum instead `guestPolicy: 'allow' | 'disallow'`.

The open/close room concept wasn't immediately obvious to me. If I understand correctly, in an open room, anybody with the `roomId` can join an ongoing call in the room (presumably only if there is already an ongoing call?). 

I can think of other interpretations of open/close for a room: whether you can add more participants once the room is created or not.

I can also think of a couple more values for the extensible `guestPolicy` enum: `allowWithPrompt` -- an organizer has to manually allow the guest in a lobby; `allowStartingCall` -- a guest can even start a call in the room as long as they have the `roomId`. Those values may not make sense right now, but an extensible enum would leave the door open for them.

elavarasidc replied:
Addressed: changed from string to extensible enum.
roomOpen is replaced by roomJoinPolicy which is an extensible enum with values ["InviteOnly", "CommunicationServiceUsers"],

prprabhu-ms commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+          "example": false
+        },
+        "participants": {
+          "description": "Collection of identities invited to the room.",
```
nit: Consistently use the term "participant" in docs.

elavarasidc replied:
Updated spec to use the term participant.

antonsamson-msft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+        "tags": [
+          "Rooms"
+        ],
+        "summary": "Retrieves an existing room by Id.",
```
nit: Do other teams use `id` or `ID` ? We have a bit of a mixture down below, e.g., `Id` and `id` as well.

elavarasidc replied:
Updated all references to id.

antonsamson-msft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+              "$ref": "#/definitions/CommunicationErrorResponse"
+            },
+            "headers": {
+              "x-ms-error-code": {
```
Do other teams document the `MS-CV` in the response headers and do we expose it in the server SDKs? Rooms service does return it in all our responses.

antonsamson-msft commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+          {
+            "in": "path",
+            "name": "roomId",
+            "description": "room id to remove the participants from.",
```
Nit: format `Room` like the others? A few more instances of lowercased Room below and some that don't have the ticks.

elavarasidc replied:
Updated all references to lowercase room to be consistent across the spec.

petrsvihlik commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+          "description": "Raw Id of the identifier. Optional in requests, required in responses.",
+          "type": "string"
+        },
+        "communicationUser": {
```
What is the reason to include this on top of the rawId?

petrsvihlik commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/examples/Rooms_AddParticipants.json:
```diff
@@ -0,0 +1,59 @@
+          "role": "Presenter"
+        },
+        {
+          "communicationIdentifier": {
+            "rawId": "8:acs:db75ed0c-e801-41a3-99a4-66a0a119a06c_00000010-ce28-064a-83fe-084822000777",
+            "communicationUser": {
+              "id": "8:acs:db75ed0c-e801-41a3-99a4-66a0a119a06c_00000010-ce28-064a-83fe-084822000777"
```
see my comment above https://github.com/Azure/azure-rest-api-specs/pull/19269/files#r897018392

elavarasidc replied:
Please see me explanation at https://github.com/Azure/azure-rest-api-specs/pull/19269/files#r898249983
No changed required here. Aligned with Chat.

petrsvihlik replied:
I'd still prefer less verbosity in the examples and remove the `rawId` considering it's optional (similarly to [callingserver](https://github.com/Azure/azure-rest-api-specs/blob/ea1a21b30050e1a4b992dcb9d33db1245fa30b89/specification/communication/data-plane/CallingServer/preview/2021-08-30-preview/examples/CallConnections_AddParticipant.json)) and is just duplicating the `communicationUser.id`. Reasoning - it requires less typing and leaves less space for error for folks who use the REST API directly without an SDK.

elavarasidc committed:
Updated spec addressing review comments from ACS Internal review

elavarasidc committed:
examples updated for roompolicy

elavarasidc committed:
Fix formatting

elavarasidc committed:
Fix example

elavarasidc commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,707 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "Request successful. The action returns a updated `Room`.",
```
Fixed.

elavarasidc commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/examples/Rooms_CreateRoom.json:
```diff
@@ -0,0 +1,62 @@
+    "api-version": "2022-02-01",
+    "createRoomRequest": {
+      "validFrom": "2022-05-30T00:00:00Z",
+      "validUntil": "2022-05-30T00:00:00Z",
```
Addressed.

elavarasidc committed:
Fix example

elavarasidc committed:
fix room

JeffreyRichter commented:
About this...
POST /rooms
We prefer PUT or PATCH /rooms/{roomId} to POST
[Response] The create room end point /rooms is a POST end point as, at the platform level we do not support customer supplied identifier. RoomID generation is controlled by IC3.

[Jeff's Response] It is not a valid reason to have an inferior API due to an IC3 implementation detail. This can easily be fixed by ACS having a table that maps the customer-provided RoomId to the IC3 ID. If you did this, the customers gets to choose IDs meaningful to them, idempotency just works (without retriabilty headers), better fault tolerance (the customer knows the ID even if they crash after getting a response). Also, you are forcing ALL your customers to keep a replicated mapping file (in a fault-tolerant way) from their friendly name to IC3's ID since they have to remember the IC3 IDs you return.

elavarasidc committed:
Role updated to RoleType as it is an enum

elavarasidc committed:
Updated participants action end points to return updated participants collection.

elavarasidc commented:
Review team, thanks for the review and feedback.  Comments have been addressed / responded.
The spec is pretty much final for rooms public preview. Please help merge this PR.
Is there any other action required from our side?

DominikMe commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,725 @@
+    "RoomModel": {
+      "description": "The meeting room.",
+      "type": "object",
+      "properties": {
```
I think all properties in this model should be marked as required.

elavarasidc replied:
Captured review comments from Dominik to the user story https://skype.visualstudio.com/SPOOL/_workitems/edit/2930042 to be worked incrementally before GA.

DominikMe commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,725 @@
+        "roomJoinPolicy": {
+          "$ref": "#/definitions/RoomJoinPolicy"
+        },
+        "participants": {
```
If we want to prepare ourselves for large collections and keep the separate /participants endpoint, we should not inline participants in the GET rooms/{roomId} response. Also to not have multiple ways of manipulating participants.

elavarasidc replied:
Captured review comments from Dominik to the user story https://skype.visualstudio.com/SPOOL/_workitems/edit/2930042 to be worked incrementally before GA.

weidongxu-microsoft reviewed (commented):
Please uncommit the "package-lock.json", the PR should not modify it.

Also there seems no action on 2 topics:
1. single id property vs. CommunicationIdentifier object
2. PUT vs. POST

elavarasidc commented:
> Please uncommit the "package-lock.json", the PR should not modify it.
> 
> Also there seems no action on 2 topics:
> 
> 1. single id property vs. CommunicationIdentifier object
> 2. PUT vs. POST

@weidongxu-microsoft, thanks for your comment. 
1. Removed package-lock.json from PR.
2. single id property vs CommunicationIdentifier Object - This was more of a discussion during board review on the communication identifier model and raw id vs mri. [No spec change].
3. PUT vs POST for create Room interface - During board review we discussed that we go with POST for public preview and will address this comment before GA. Tracking this on a work item tagged pre-GA.

elavarasidc committed:
Adding security.

elavarasidc commented on specification/communication/data-plane/Rooms/preview/2022-02-01-preview/communicationservicesrooms.json:
```diff
@@ -0,0 +1,725 @@
+  },
+  "securityDefinitions": {
+    "Authorization": {
+      "type": "apiKey",
```
Yes, authentication with AAD is not enabled yet for Rooms.
Added security property. Thanks.

elavarasidc committed:
Fixing swagger lint error

weidongxu-microsoft commented:
/azp run

weidongxu-microsoft commented:
Failure on `ApiDocPreview` appears not related to the PR.

As this is not a required CI, it will not block PR merge.

Error message
```
##[error]Run Git command -c http.extraHeader="Authorization: Basic ***" clone https://github.com/Azure/azure-rest-api-specs _swagger_specs --depth 1 --shallow-submodules --no-single-branch failed. Error: Cloning into '_swagger_specs'... Updating files:   2% (2188/91270) Updating files:   3% (2739/91270) Updating files:   4% (3651/91270) Updating files:   5% (4564/91270) Updating files:   5% (5040/91270) Updating files:   6% (5477/91270) Updating files:   7% (6389/91270) Updating files:   7% (6704/91270) Updating files:   8% (7302/91270) Updating files:   9% (8215/91270) Updating files:   9% (8328/91270) Updating files:  10% (9127/91270) Updating files:  11% (10040/91270) Updating files:  11% (10419/91270) Updating files:  12% (10953/91270) Updating files:  13% (11866/91270) Updating files:  14% (12778/91270) Updating files:  14% (12879/91270) Updating files:  15% (13691/91270) Updating files:  15% (14330/91270) Updating files:  16% (14604/91270) Updating files:  17% (15516/91270) Updating files:  18% (16429/91270) Updating files:  19% (17342/91270) Updating files:  19% (17589/91270) error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineExamples/VirtualMachine_Create_WindowsVmWithPatchSettingModeOfAutomaticByPlatformAndEnableHotPatchingTrue.json: Filename too long error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineScaleSetExamples/VirtualMachineScaleSets_ForceRecoveryServiceFabricPlatformUpdateDomainWalk_MaximumSet_Gen.json: Filename too long error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineScaleSetExamples/VirtualMachineScaleSets_ForceRecoveryServiceFabricPlatformUpdateDomainWalk_MinimumSet_Gen.json: Filename too long Updating files:  20% (18254/91270) Updating files:  21% (19167/91270) Updating files:  21% (19561/91270) Updating files:  22% (20080/91270) Updating files:  22% (20726/91270) Updating files:  23% (20993/91270) Updating files:  24% (21905/91270) Updating files:  24% (21975/91270) Updating files:  25% (22818/91270) Updating files:  26% (23731/91270) Updating files:  27% (24643/91270) Updating files:  27% (24925/91270) Updating files:  28% (25556/91270) Updating files:  29% (26469/91270) Updating files:  29% (26626/91270) Updating files:  30% (27381/91270) Updating files:  30% (27767/91270) Updating files:  31% (28294/91270) Updating files:  31% (29096/91270) Updating files:  32% (29207/91270) Updating files:  33% (30120/91270) Updating files:  33% (30712/91270) Updating files:  34% (31032/91270) Updating files:  35% (31945/91270) Updating files:  35% (32204/91270) Updating files:  36% (32858/91270) Updating files:  36% (33309/91270) Updating files:  37% (33770/91270) Updating files:  37% (34327/91270) Updating files:  38% (34683/91270) Updating files:  39% (35596/91270) Updating files:  39% (35706/91270) Updating files:  40% (36508/91270) Updating files:  40% (36885/91270) Updating files:  41% (37421/91270) Updating files:  42% (38334/91270) Updating files:  42% (38939/91270) Updating files:  43% (39247/91270) Updating files:  43% (40145/91270) Updating files:  44% (40159/91270) Updating files:  45% (41072/91270) Updating files:  45% (41347/91270) Updating files:  46% (41985/91270) Updating files:  46% (42198/91270) Updating files:  47% (42897/91270) Updating files:  47% (43796/91270) Updating files:  48% (43810/91270) Updating files:  49% (44723/91270) Updating files:  49% (45439/91270) Updating files:  50% (45635/91270) Updating files:  51% (46548/91270) Updating files:  51% (47022/91270) Updating files:  52% (47461/91270) Updating files:  53% (48374/91270) Updating files:  53% (48511/91270) Updating files:  53% (48689/91270) Updating files:  54% (49286/91270) Updating files:  55% (50199/91270) Updating files:  55% (50603/91270) Updating files:  56% (51112/91270) Updating files:  57% (52024/91270) Updating files:  57% (52643/91270) Updating files:  58% (52937/91270) Updating files:  59% (53850/91270) Updating files:  60% (54762/91270) Updating files:  60% (55506/91270) Updating files:  61% (55675/91270) Updating files:  62% (56588/91270) Updating files:  63% (57501/91270) Updating files:  64% (58413/91270) Updating files:  64% (58559/91270) Updating files:  65% (59326/91270) Updating files:  66% (60239/91270) Updating files:  67% (61151/91270) Updating files:  67% (61238/91270) Updating files:  68% (62064/91270) Updating files:  68% (62144/91270) Updating files:  68% (62731/91270) Updating files:  69% (62977/91270) Updating files:  70% (63889/91270) Updating files:  71% (64802/91270) Updating files:  71% (65090/91270) Updating files:  72% (65715/91270) Updating files:  72% (66307/91270) Updating files:  73% (66628/91270) Updating files:  74% (67540/91270) Updating files:  74% (67570/91270) Updating files:  75% (68453/91270) Updating files:  75% (68599/91270) Updating files:  76% (69366/91270) Updating files:  76% (69673/91270) Updating files:  77% (70278/91270) Updating files:  77% (70870/91270) Updating files:  78% (71191/91270) Updating files:  79% (72104/91270) Updating files:  79% (72583/91270) Updating files:  80% (73016/91270) Updating files:  81% (73929/91270) Updating files:  81% (73939/91270) Updating files:  82% (74842/91270) Updating files:  83% (75755/91270) Updating files:  83% (76272/91270) Updating files:  84% (76667/91270) Updating files:  85% (77580/91270) Updating files:  85% (77768/91270) Updating files:  86% (78493/91270) Updating files:  86% (79104/91270) Updating files:  87% (79405/91270) Updating files:  88% (80318/91270) Updating files:  89% (81231/91270) Updating files:  89% (81733/91270) Updating files:  90% (82143/91270) Updating files:  91% (83056/91270) Updating files:  91% (83308/91270) Updating files:  92% (83969/91270) Updating files:  92% (84309/91270) Updating files:  93% (84882/91270) Updating files:  93% (85440/91270) Updating files:  94% (85794/91270) Updating files:  95% (86707/91270) Updating files:  95% (87224/91270) Updating files:  96% (87620/91270) Updating files:  96% (88489/91270) Updating files:  97% (88532/91270) Updating files:  98% (89445/91270) Updating files:  98% (89790/91270) Updating files:  99% (90358/91270) Updating files: 100% (91270/91270) Updating files: 100% (91270/91270), done. fatal: unable to checkout working tree warning: Clone succeeded, but checkout failed. You can inspect what was checked out with 'git status' and retry with 'git restore --source=HEAD :/' 
Run Git command -c http.extraHeader="Authorization: Basic ***" clone https://github.com/Azure/azure-rest-api-specs _swagger_specs --depth 1 --shallow-submodules --no-single-branch failed. Error: Cloning into '_swagger_specs'... Updating files:   2% (2188/91270) Updating files:   3% (2739/91270) Updating files:   4% (3651/91270) Updating files:   5% (4564/91270) Updating files:   5% (5040/91270) Updating files:   6% (5477/91270) Updating files:   7% (6389/91270) Updating files:   7% (6704/91270) Updating files:   8% (7302/91270) Updating files:   9% (8215/91270) Updating files:   9% (8328/91270) Updating files:  10% (9127/91270) Updating files:  11% (10040/91270) Updating files:  11% (10419/91270) Updating files:  12% (10953/91270) Updating files:  13% (11866/91270) Updating files:  14% (12778/91270) Updating files:  14% (12879/91270) Updating files:  15% (13691/91270) Updating files:  15% (14330/91270) Updating files:  16% (14604/91270) Updating files:  17% (15516/91270) Updating files:  18% (16429/91270) Updating files:  19% (17342/91270) Updating files:  19% (17589/91270) error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineExamples/VirtualMachine_Create_WindowsVmWithPatchSettingModeOfAutomaticByPlatformAndEnableHotPatchingTrue.json: Filename too long error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineScaleSetExamples/VirtualMachineScaleSets_ForceRecoveryServiceFabricPlatformUpdateDomainWalk_MaximumSet_Gen.json: Filename too long error: unable to create file specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/ComputeRP/examples/virtualMachineScaleSetExamples/VirtualMachineScaleSets_ForceRecoveryServiceFabricPlatformUpdateDomainWalk_MinimumSet_Gen.json: Filename too long Updating files:  20% (18254/91270) Updating files:  21% (19167/91270) Updating files:  21% (19561/91270) Updating files:  22% (20080/91270) Updating files:  22% (20726/91270) Updating files:  23% (20993/91270) Updating files:  24% (21905/91270) Updating files:  24% (21975/91270) Updating files:  25% (22818/91270) Updating files:  26% (23731/91270) Updating files:  27% (24643/91270) Updating files:  27% (24925/91270) Updating files:  28% (25556/91270) Updating files:  29% (26469/91270) Updating files:  29% (26626/91270) Updating files:  30% (27381/91270) Updating files:  30% (27767/91270) Updating files:  31% (28294/91270) Updating files:  31% (29096/91270) Updating files:  32% (29207/91270) Updating files:  33% (30120/91270) Updating files:  33% (30712/91270) Updating files:  34% (31032/91270) Updating files:  35% (31945/91270) Updating files:  35% (32204/91270) Updating files:  36% (32858/91270) Updating files:  36% (33309/91270) Updating files:  37% (33770/91270) Updating files:  37% (34327/91270) Updating files:  38% (34683/91270) Updating files:  39% (35596/91270) Updating files:  39% (35706/91270) Updating files:  40% (36508/91270) Updating files:  40% (36885/91270) Updating files:  41% (37421/91270) Updating files:  42% (38334/91270) Updating files:  42% (38939/91270) Updating files:  43% (39247/91270) Updating files:  43% (40145/91270) Updating files:  44% (40159/91270) Updating files:  45% (41072/91270) Updating files:  45% (41347/91270) Updating files:  46% (41985/91270) Updating files:  46% (42198/91270) Updating files:  47% (42897/91270) Updating files:  47% (43796/91270) Updating files:  48% (43810/91270) Updating files:  49% (44723/91270) Updating files:  49% (45439/91270) Updating files:  50% (45635/91270) Updating files:  51% (46548/91270) Updating files:  51% (47022/91270) Updating files:  52% (47461/91270) Updating files:  53% (48374/91270) Updating files:  53% (48511/91270) Updating files:  53% (48689/91270) Updating files:  54% (49286/91270) Updating files:  55% (50199/91270) Updating files:  55% (50603/91270) Updating files:  56% (51112/91270) Updating files:  57% (52024/91270) Updating files:  57% (52643/91270) Updating files:  58% (52937/91270) Updating files:  59% (53850/91270) Updating files:  60% (54762/91270) Updating files:  60% (55506/91270) Updating files:  61% (55675/91270) Updating files:  62% (56588/91270) Updating files:  63% (57501/91270) Updating files:  64% (58413/91270) Updating files:  64% (58559/91270) Updating files:  65% (59326/91270) Updating files:  66% (60239/91270) Updating files:  67% (61151/91270) Updating files:  67% (61238/91270) Updating files:  68% (62064/91270) Updating files:  68% (62144/91270) Updating files:  68% (62731/91270) Updating files:  69% (62977/91270) Updating files:  70% (63889/91270) Updating files:  71% (64802/91270) Updating files:  71% (65090/91270) Updating files:  72% (65715/91270) Updating files:  72% (66307/91270) Updating files:  73% (66628/91270) Updating files:  74% (67540/91270) Updating files:  74% (67570/91270) Updating files:  75% (68453/91270) Updating files:  75% (68599/91270) Updating files:  76% (69366/91270) Updating files:  76% (69673/91270) Updating files:  77% (70278/91270) Updating files:  77% (70870/91270) Updating files:  78% (71191/91270) Updating files:  79% (72104/91270) Updating files:  79% (72583/91270) Updating files:  80% (73016/91270) Updating files:  81% (73929/91270) Updating files:  81% (73939/91270) Updating files:  82% (74842/91270) Updating files:  83% (75755/91270) Updating files:  83% (76272/91270) Updating files:  84% (76667/91270) Updating files:  85% (77580/91270) Updating files:  85% (77768/91270) Updating files:  86% (78493/91270) Updating files:  86% (79104/91270) Updating files:  87% (79405/91270) Updating files:  88% (80318/91270) Updating files:  89% (81231/91270) Updating files:  89% (81733/91270) Updating files:  90% (82143/91270) Updating files:  91% (83056/91270) Updating files:  91% (83308/91270) Updating files:  92% (83969/91270) Updating files:  92% (84309/91270) Updating files:  93% (84882/91270) Updating files:  93% (85440/91270) Updating files:  94% (85794/91270) Updating files:  95% (86707/91270) Updating files:  95% (87224/91270) Updating files:  96% (87620/91270) Updating files:  96% (88489/91270) Updating files:  97% (88532/91270) Updating files:  98% (89445/91270) Updating files:  98% (89790/91270) Updating files:  99% (90358/91270) Updating files: 100% (91270/91270) Updating files: 100% (91270/91270), done. fatal: unable to checkout working tree warning: Clone succeeded, but checkout failed. You can inspect what was checked out with 'git status' and retry with 'git restore --source=HEAD :/' 
--- Failed exec run.ps1 with exception Run.ps1 failed with exit code 128 
```

elavarasidc commented:
Team, please help merge this PR.

ruowan commented:
/azp run
